### PR TITLE
Line labels code refactored as plug-in, using the plug-in bar

### PR DESCRIPTION
**Pull request by ibusko**
This PR replaces the old one (#553) which kept the plug-in in a floating, separate window. In this one, the line labels controls are all kept confined inside the plug-in bar.

### CODE REVIEW

**Comment by ibusko** [comment]
can't understand the build errors though.

**Comment by ibusko** [comment]
There are 4 configs that error with this cryptic message. I couldn't figure out what is causing them:

PyQt5 5.11.3 -- Qt runtime 5.11.2 -- Qt compiled 5.11.2
rootdir: /private/var/folders/nz/vv4_9tw56nv9k3tkvyszvwg80000gn/T/specviz-test-z667r7ij/lib/python3.6/site-packages, inifile: setup.cfg
plugins: remotedata-0.3.1, qt-3.2.2, openfiles-0.3.2, doctestplus-0.2.0, arraydiff-0.3
collecting ... The command "$MAIN_CMD $SETUP_CMD" exited with 250.

**Comment by nmearl** [comment]
@ibusko I'm looking into this today and am seeing the say thing even locally. I'm certain it's because of some issues with the qtbot tests.

**Comment by nluetzge** [comment]
Played around with it. Seems to work fine with scaling, and unit change. I had problems finding the line lists tho. Can we make it that it offers the available line lists in the plugin again how it was before? Also is it possible to name newly created sets? At the moment they are just called 1, 2, 3 and this is can get a bit confusing.

**Comment by ibusko** [comment]
@nluetzge do you see the internal line lists in the drop-down that says "Select line list" in the plug-in? Or is it empty? I notice that it sometimes flashes when its selector button (the up/down arrow at its right corner) is clicked once. When I click a second time I get the fully populated drop-down. This happens in the older plug-in as well (the one in master).

**Comment by nluetzge** [comment]
@ibusko oh this is so weird. I see it now. I swear it was not there earlier... (the whole drop down menu I mean) but I also didn't update anything. Let me see if I can reproduce this.

**Comment by nluetzge** [comment]
Ahh! Its not there if the window is too small!

**Comment by ibusko** [comment]
@nluetzge it looks like this is a kind of generic issue with the plug-in bar: if the user screen is too small to fit the plug-in GUI, it will get some elements hidden when first depicted on screen. I recall struggling with layout managers when converting the plug-in from a floating window to a pinned window in the plug-in bar. This was the best I could achieve without a major redesign of the entire GUI. Maybe we should address this as a separate issue in thfuture.

**Comment by ibusko** [comment]
@nmearl these two conficting files are files that existed before but were moved to the plugin subdirectory. Both core/ and widgets/ directories are now free of any line labels-specific stuff. I don't know why they are generating a conflict. Should the fix be made in master or in the branch? The branch no longer has these files anyway.

**Comment by nmearl** [comment]
@ibusko You need to rebase the branch in fork onto master, that should remove the issue. If you want, you could even pull from my [fork of your branch](https://github.com/nmearl/specviz/tree/ivo-line-labels) that I've been using to test pytest.

**Comment by ibusko** [comment]
@nmearl I rebased my github fork, but still the astropy_helpers directory refuses to merge. There seems to be a version problem. Do you know how to fix this? I don't want to commit something that I didn't actually change. This is the message I get from diff'ing the directory:

(specviz_dev) kamino:/Users/busko $ git diff astropy_helpers
diff --git a/astropy_helpers b/astropy_helpers
--- a/astropy_helpers
+++ b/astropy_helpers
@@ -1 +1 @@
-Subproject commit 35b16cd0baa45e6c58d1d4e48b80034b66857d1c
+Subproject commit 35b16cd0baa45e6c58d1d4e48b80034b66857d1c-dirty
(specviz_dev) kamino:/Users/busko $

**Comment by nmearl** [comment]
You should be able to simple reset the `astropy_helpers` directory to what's in master by running
`git checkout upstream/master -- astropy_helpers`. This should bring it up-to-date with what's currently in master.

**Comment by ibusko** [comment]
@nmearl OK, now it's clear of conflicts. Thanks for the hints!

**Comment by nmearl** [comment]
@ibusko I'm still seeing a lot merge commits in the history of this PR -- did you rebase the entire thing?

**Comment by ibusko** [comment]
@nmearl I think I rebased the entire branch. Yes, it's weird why so many commits still show up.

Btw, I commited one last change, adding doc strings to some top classes in the line labels plugin. By mistake, the commit went with the wrong message (Interim).

File: specviz/app.py [hunk 1]
```diff
@@ -42,7 +42,7 @@ def __init__(self, *args, file_path=None, file_loader=None, embedded=False,
         self.load_local_plugins()
 
         # cache the line lists for speedier access
-        from .core import linelist
+        from .plugins.line_labels import linelist
```
Comment (nmearl) [inline specviz/app.py]: This should not be explicitly imported. Plugins defined with the decorator will automatically be imported into the namespace.

File: specviz/plugins/line_labels/line_labels_plotter.py [hunk 2]
```diff
@@ -406,13 +468,6 @@ def _destroy_zoom_markers_thread(self):
 
 class ZoomMarkersThread(QThread):
```
Comment (nmearl) [inline specviz/plugins/line_labels/line_labels_plotter.py]: This class will need a simple docstring.

File: specviz/widgets/plotting.py [hunk 3]
```diff
@@ -169,7 +163,7 @@ def __init__(self, title=None, model=None, visible=True, *args, **kwargs):
         self._visible = visible
 
         # Performance enhancements
-        # self.setDownsampling(auto=False)
+        self.setDownsampling(auto=True)
```
Comment (nmearl) [inline specviz/widgets/plotting.py]: This should remain commented out.

File: specviz/app.py (hunk 1 above)
Comment (ibusko) [reply specviz/app.py]: Done.

File: specviz/plugins/line_labels/line_labels_plotter.py (hunk 2 above)
Comment (ibusko) [reply specviz/plugins/line_labels/line_labels_plotter.py]: Done.

File: specviz/widgets/plotting.py (hunk 3 above)
Comment (ibusko) [reply specviz/widgets/plotting.py]: Done.

**Comment by ibusko** [comment]
@nmearl I put in the modifications. The only one I'm struggling with is with astropy_helpers. I don't know why it was modified in the first place, and a diff doesn't show any differences with master.

When I try to pull in a fresh checkout from upstream, I get an error:

bash-3.2$ git remote -v
origin	https://github.com/ibusko/specviz.git (fetch)
origin	https://github.com/ibusko/specviz.git (push)
upstream	https://github.com/spacetelescope/specviz (fetch)
upstream	https://github.com/spacetelescope/specviz (push)
bash-3.2$ git checkout upstream/master -- astropy_helpers
fatal: invalid reference: upstream/master
bash-3.2$_

**Comment by nmearl** [comment]
Have you done a `git fetch upstream` lately?

**Comment by ibusko** [comment]
@nmearl no, I hadn't. That fixed the checkout from upstream, but I couldn't commit from within pycharm. It commited and got pushed from the command line though. Something definitely got bonkers in my environmemt. After this version is released, I probably will start a new github fork from scratch.

**Comment by nmearl** [comment]
If you hadn't done a `fetch` lately, you may want to `rebase` again too. It may get rid of the large numbers of merge commits in your pr history.

**Comment by ibusko** [comment]
@nmearl although the PR is no longer reporting merge conflicts, the rebase / squash I am playing with reports conflicts and is not successful in squashing any commits.  If the excessive number of commits is not a blocker, I would rather leave it as is, at least for now.

**Comment by nmearl** [comment]
It's not the number of commits -- don't squash anything -- it's the number of `Merge remote-tracking branch 'origin/plugin_linelabels' into plugin...` commits. These should disappear on a rebase.

**Comment by nmearl** [comment]
If you look at the commit history of my branch of your fork [here](https://github.com/nmearl/specviz/commits/ivo-line-labels), you'll see that they are all just your code commits, without the merge stuff in it.

**Comment by nmearl** [comment]
You will have to resolve the conflicts, yes. PyCharm can help with this if, after you do a `git rebase upstream/master`, you go to VCS -> Git -> Resolve Conflicts. Once you accept the changes, go to VCS -> Git -> Continue Rebase. It will give you a nice UI for how to resolve the conflicts.

**Comment by ibusko** [comment]
Yes, I did it already. The problem is that rebase returns with an error:

Applying: More debugging.
Using index info to reconstruct a base tree...
M	specviz/core/annotation.py
Falling back to patching base and 3-way merge...
Removing specviz/core/linelist.py
CONFLICT (modify/delete): specviz/core/annotation.py deleted in More debugging. and modified in HEAD. Version HEAD of specviz/core/annotation.py left in tree.
Patch failed at 0016 More debugging.
Use 'git am --show-current-patch' to see the failed patch
Resolve all conflicts manually, mark them as resolved with
"git add/rm <conflicted_files>", then run "git rebase --continue".
You can instead skip this commit: run "git rebase --skip".
To abort and get back to the state before "git rebase", run "git rebase --abort".
error: Failed to merge in the changes.

I tried the --skip option, it skips this commit but stops a few commits further down with another conflict. And goes like that for dozens of commits. Eventually ends successfully, but the resulting code is broken. I can't skip too many commits, I guess. So I though maybe squashing them, just the ones with conflicts, could fix it, but I'm not sure. 

Note that I had created a new branch before so as to preserve the working code on github. I made another PR for that new branch. Don't know how to proceed from here.

**Comment by ibusko** [comment]
I see what you mean. I get one of these errors for every conflict in the commit history. Isn't there a way to batch-process all of them? With PyCharm I can only do one by one, and my standard action is "Accept yours".

**Comment by ibusko** [comment]
@nmearl eventually it got to the end. The new PR doesn't have the spurious commits anymore. So maybe we should move our attention to the new PR (#646) and keep this here closed.

**Pull request closed without merging**